Support the case where type mapper is a class

YutaUra opened the pull request:
Hi @eddeee888 .
Sorry for the sudden creation of this PR.

Currently, Type Mapper correctly generates missing resolvers only for Type or Interface, but when Type Mapper is a class, it recognizes all fields as missing resolvers.

Below is a specific example

```graphql
# schema.graphql
type User {
  id: ID!
  name: String!
}
```

```typescript
// schema.mappers.ts

export class UserMapper {
  id: string
  name: string
}
```

```typescript
// resolvers/User.ts
import type { UserResolvers } from './types.generated';

export const User: UserResolvers = {
  /* Implement logic here */
  id: () => { /* User.id resolver is required because User.firstName exists but UserMapper.id does not */ },
  name: () => { /* User.name resolver is required because User.name exists but UserMapper.id does not */ },
  // If there were more fields, it would generate a lot of them...
};
```

I find this PR useful, at least for me, but I guess there is a reason why it has not been implemented so far, on the other hand?

Perhaps the scope of what I am able to consider may be narrower than I imagine. If so, I would very much appreciate your advice and feedback.

eddeee888 commented:
This is great @YutaUra ! I'll take a look this coming week

eddeee888 commented on packages/typescript-resolver-files/src/utils/getNodePropertyMap.ts:
```diff
@@ -1,3 +1,4 @@
+import { ClassDeclaration } from 'ts-morph';
 import { type TypeNode, Node, SyntaxKind } from 'ts-morph';
```
Nitpicky: I think this can be merged 🙂

```
import { type TypeNode, type ClassDeclaration, Node, SyntaxKind } from 'ts-morph';
```

eddeee888 commented on packages/typescript-resolver-files/src/utils/getNodePropertyMap.ts:
```diff
@@ -63,3 +68,31 @@ const collectTypeNodeProperties = (
   }
 };
+
+const collectClassNodeProperties = (
```
Could we add a unit test for this please? 🙂

eddeee888 commented:
Thanks @YutaUra ! This is looking great!

Do you mind adding an e2e test for this case, please?

Here's what I think should work:

1. Add the following setup to `test-modules` (the main happy path):

```graphql
# packages/typescript-resolver-files-e2e/src/test-modules/modules/book/book.graphqls
extend type Query {
  book(id: ID!): Book
}

type Book {
  id: ID!
  isbn: String!
  author: User!
  mainGenre: String!
  publishedAt: DateTime!
}
```

```ts
// packages/typescript-resolver-files-e2e/src/test-modules/modules/book/book.mappers.ts
export type BookMapper = {
  id: string;
  isbn: string;
  genre: string[];
  authorUserId: string;
  publishedAt: string;
};
```

2. Run codegen on this test: `yarn nx graphql-codegen typescript-resolver-files-e2e -c test-modules`
3. Confirm the following generated file templates are as expected:
- `packages/typescript-resolver-files-e2e/src/test-modules/modules/book/resolvers/Query/book.ts`
- `packages/typescript-resolver-files-e2e/src/test-modules/modules/book/resolvers/Book.ts`

4. Commit the e2e test setup and generated files.

Thank you!

YutaUra commented:
Thanks for the review too! @eddeee888

I've fixed the imports (which I totally missed) and added an e2e test, what do you think?

---

And this part should be a class, right?
In the meantime, I did it with a class and it seemed to work fine, so please check that too!

> ```
> // packages/typescript-resolver-files-e2e/src/test-modules/modules/book/book.mappers.ts
> export type BookMapper = { // <- This should be class?
>   id: string;
>   isbn: string;
>   genre: string[];
>   authorUserId: string;
>   publishedAt: string;
> };
> ```

eddeee888 reviewed (approved):
Thanks so much for this improvement! 🙌
